File: Python3/427. Construct Quad Tree.py
"""
# Definition for a QuadTree node.
class Node:
    def __init__(self, val, isLeaf, topLeft, topRight, bottomLeft, bottomRight):
        self.val = val
        self.isLeaf = isLeaf
        self.topLeft = topLeft
        self.topRight = topRight
        self.bottomLeft = bottomLeft
        self.bottomRight = bottomRight
"""

class Solution:
    def construct(self, grid: List[List[int]]) -> 'Node':

        def helper(x: int, y: int, size: int) -> 'Node':
            if size == 1:
                return Node(grid[x][y], True)

            s = size // 2
            topLeft = helper(x, y, s)
            topRight = helper(x, y + s, s)
            bottomLeft = helper(x + s, y, s)
            bottomRight = helper(x + s, y + s, s)

            if topLeft.isLeaf and topRight.isLeaf and bottomLeft.isLeaf and bottomRight.isLeaf \
            and topLeft.val == topRight.val \
            and topRight.val == bottomLeft.val \
            and bottomLeft.val == bottomRight.val:
                return Node(topLeft.val, True)

            return Node(False, False, topLeft, topRight, bottomLeft, bottomRight)

        return helper(0, 0, len(grid))

File: Java/427. Construct Quad Tree.java
/*
// Definition for a QuadTree node.
class Node {
    public boolean val;
    public boolean isLeaf;
    public Node topLeft;
    public Node topRight;
    public Node bottomLeft;
    public Node bottomRight;

    
    public Node() {
        this.val = false;
        this.isLeaf = false;
        this.topLeft = null;
        this.topRight = null;
        this.bottomLeft = null;
        this.bottomRight = null;
    }
    
    public Node(boolean val, boolean isLeaf) {
        this.val = val;
        this.isLeaf = isLeaf;
        this.topLeft = null;
        this.topRight = null;
        this.bottomLeft = null;
        this.bottomRight = null;
    }
    
    public Node(boolean val, boolean isLeaf, Node topLeft, Node topRight, Node bottomLeft, Node bottomRight) {
        this.val = val;
        this.isLeaf = isLeaf;
        this.topLeft = topLeft;
        this.topRight = topRight;
        this.bottomLeft = bottomLeft;
        this.bottomRight = bottomRight;
    }
};
*/

class Solution {
    public Node construct(int[][] grid) {
        return helper(grid, 0, 0, grid.length);
    }

    private Node helper(int[][] grid, int x, int y, int size) {
        // reach to a leaf node
        if (size == 1) {
            return new Node(grid[x][y] == 1, true);
        }

        size /= 2;
        Node topLeft = helper(grid, x, y, size);
        Node topRight = helper(grid, x, y + size, size);
        Node bottomLeft = helper(grid, x + size, y, size);
        Node bottomRight = helper(grid, x + size, y + size, size);

        if (topLeft.isLeaf && topRight.isLeaf && bottomLeft.isLeaf && bottomRight.isLeaf &&
            topLeft.val == topRight.val && topRight.val == bottomLeft.val && 
            bottomLeft.val == bottomRight.val) {
                return new Node(topLeft.val, true);
        }
        return new Node(false, false, topLeft, topRight, bottomLeft, bottomRight);
    }
}